lib/spack/spack/environment.py

```diff
@@ -540,36 +604,39 @@ def regenerate(self, all_specs, roots):
-            )
-            fs.mkdirp(root)
-
-            # The tempdir for the directory transaction must be in the same
-            # filesystem mount as the view for symlinks to work. Provide
-            # dirname(root) as the tempdir for the
-            # replace_directory_transaction because it must be on the same
-            # filesystem mount as the view itself. Otherwise it may be
-            # impossible to construct the view in the tempdir even when it can
-            # be constructed in-place.
-            with fs.replace_directory_transaction(root, os.path.dirname(root)):
-                view = self.view()
-
-                view.clean()
-                specs_in_view = set(view.get_all_specs())
-                tty.msg("Updating view at {0}".format(self.root))
-
-                rm_specs = specs_in_view - installed_specs_for_view
-                add_specs = installed_specs_for_view - specs_in_view
-
-                # pass all_specs in, as it's expensive to read all the
-                # spec.yaml files twice.
-                view.remove_specs(*rm_specs, with_dependents=False,
-                                  all_specs=specs_in_view)
-                view.add_specs(*add_specs, with_dependencies=False)
+            # we always regenerate the view from scratch.
+            # We will do this by hashing the view contents and putting the view
+            # in a directory by hash, and then having a symlink to the real
+            # view in the root. The real root for a view at /dirname/basename
+            # will be /dirname/._basename_<hash>.
+            # This allows for atomic swaps when we update the view
+            # construct view at new_root
+            tty.msg("Updating view at {0}".format(self.root))
+
+            # cache the roots because the way we determine which is which does
+            # not work while we are updating
+            new_root = self._next_root(installed_specs_for_view)
+            old_root = self._current_root
+
+            if new_root != old_root:  # if equal, we're constructing same view
+                view = self.view(new=new_root)
+                fs.mkdirp(new_root)
+                view.add_specs(*installed_specs_for_view,
+                               with_dependencies=False)
+
+                # create symlink from tmpname to new_root
+                root_dirname = os.path.dirname(self.root)
+                tmp_symlink_name = os.path.join(root_dirname, '._view_link')
+                os.symlink(new_root, tmp_symlink_name)
+
+                # mv symlink atomically over root symlink to old_root
+                if os.path.exists(self.root) and not os.path.islink(self.root):
+                    raise RuntimeError("Cannot create view: file already exists")
+                os.rename(tmp_symlink_name, self.root)
+
+                # remove old_root
+                if old_root and os.path.exists(old_root):
+                    shutil.rmtree(old_root)
```

Comment:
does this fail the whole operation if it fails?  I think it should warn since we don't actually care about the old path if rename succeeds.